### PR TITLE
Adds handling for when regions are unreachable

### PR DESCRIPTION
**Pull request by joehan**
### Description
Added handling for when regions are unreachable during functions deploy and functions:delete.

### Scenarios Tested
Tested out that a deploy still works if no regions are unreachable. Haven't been able to test the error cases though.

### CODE REVIEW

File: src/deploy/functions/deploy.ts [hunk 1]
```diff
@@ -26,9 +26,6 @@ async function uploadSource(context: any): Promise<void> {
 // eslint-disable-next-line @typescript-eslint/no-explicit-any
 export async function deploy(context: any, options: any, payload: any): Promise<void> {
   if (options.config.get("functions")) {
-    context.existingFunctions =
```
Comment (joehan) [inline src/deploy/functions/deploy.ts:29]: Moved this earlier in the process.

File: src/deploy/functions/prepare.ts [hunk 2]
```diff
@@ -87,5 +89,28 @@ export async function prepare(context: any, options: any, payload: any): Promise
...
+  if (res.unreachable) {
+    const regionsInDeployment = Object.keys(payload.functions.byRegion);
+    const unreachableRegionsInDeployment = regionsInDeployment.filter((region) =>
+      res.unreachable.includes(region)
+    );
+    if (unreachableRegionsInDeployment) {
+      throw new FirebaseError(
+        "The following Cloud Functions regions are currently unreachable:\n" +
+          unreachableRegionsInDeployment.join("\n") +
+          "This deployment contains functions in those regions. Please try again in a few minutes, or exclude these regions from your deployment."
```
Comment (taeold) [inline src/deploy/functions/prepare.ts:103]: nit* I think we might want to add a space or a new line before the last sentence for readability. Similar changes to other log lines.

**Review by taeold — APPROVED** [review]

File: src/deploy/functions/prepare.ts (hunk 2 above)
Comment (joehan) [reply src/deploy/functions/prepare.ts:103]: Ahhh, good point. Will do

File: src/deploy/functions/prompts.ts [hunk 3]
```diff
@@ -30,7 +30,7 @@ export async function promptForFailurePolicies(
 
   context.existingFunctions =
     (context.existingFunctions as CloudFunctionTrigger[]) ||
-    (await gcp.cloudfunctions.listAllFunctions(context.projectId));
+    (await gcp.cloudfunctions.listAllFunctions(context.projectId)).functions;
```
Comment (inlined) [inline src/deploy/functions/prompts.ts:33]: In prepare.ts you only stored the functions.

File: src/functionsDeployHelper.ts [hunk 4]
```diff
@@ -211,8 +211,8 @@ export function printSuccess(funcName: string, type: string) {
 }
 
 export async function printTriggerUrls(projectId: string, sourceUrl: string) {
-  const functions = await cloudfunctions.listAllFunctions(projectId);
-  const httpsFunctions = functions.filter((fn) => {
+  const res = await cloudfunctions.listAllFunctions(projectId);
```
Comment (inlined) [inline src/functionsDeployHelper.ts:214]: Separately, why is this needed? We already have the results from create and update.

File: src/gcp/cloudfunctions.ts [hunk 5]
```diff
@@ -435,17 +441,17 @@ export async function listFunctions(projectId: string, region: string): Promise<
       origin: api.functionsOrigin,
     });
     if (res.body.unreachable && res.body.unreachable.length > 0) {
-      throw new FirebaseError(
-        "Some Cloud Functions regions were unreachable, please try again later.",
-        { exit: 2 }
-      );
+      logger.debug(`[functions] unable to reach the following regions: ${res.body.unreachable}`);
```
Comment (inlined) [inline src/gcp/cloudfunctions.ts:444]: `unreachable` is a list. Let's join it for debug printing.

File: src/gcp/cloudfunctions.ts [hunk 6]
```diff
@@ -426,7 +426,13 @@ export async function deleteFunction(options: any): Promise<Operation> {
  * @param projectId the Id of the project to check.
  * @param region the region to check in.
  */
-export async function listFunctions(projectId: string, region: string): Promise<CloudFunction[]> {
+export async function listFunctions(
```
Comment (inlined) [inline src/gcp/cloudfunctions.ts:429]: I don't see how it's possible to get unreachable regions when region isn't `-`. Should we use a common helper method and then make `listFunctions` return just a `CloudFunction[]`?

File: src/deploy/functions/prepare.ts [hunk 7]
```diff
@@ -87,5 +89,28 @@ export async function prepare(context: any, options: any, payload: any): Promise
...
+  const res = await gcp.cloudfunctions.listAllFunctions(context.projectId);
+  if (res.unreachable) {
+    const regionsInDeployment = Object.keys(payload.functions.byRegion);
+    const unreachableRegionsInDeployment = regionsInDeployment.filter((region) =>
+      res.unreachable.includes(region)
+    );
+    if (unreachableRegionsInDeployment) {
+      throw new FirebaseError(
+        "The following Cloud Functions regions are currently unreachable:\n" +
+          unreachableRegionsInDeployment.join("\n") +
```
Comment (inlined) [inline src/deploy/functions/prepare.ts:102]: nit: Wouldn't it be more consistent in the style to put a tab before elements in a list?

File: src/gcp/cloudfunctions.ts [hunk 8]
```diff
@@ -457,7 +463,12 @@ export async function listFunctions(projectId: string, region: string): Promise<
  * List all existing Cloud Functions in a project.
  * @param projectId the Id of the project to check.
  */
-export async function listAllFunctions(projectId: string): Promise<CloudFunction[]> {
+export async function listAllFunctions(
+  projectId: string
+): Promise<{
```
Comment (inlined) [inline src/gcp/cloudfunctions.ts:468]: Nit: consider making this a named type so tests can declare things to use that type.

File: src/functionsDeployHelper.ts (hunk 4 above)
Comment (joehan) [reply src/functionsDeployHelper.ts:214]: Good point, this is probably an unnecessary call.

File: src/deploy/functions/prompts.ts (hunk 3 above)
Comment (joehan) [reply src/deploy/functions/prompts.ts:33]: Looking closer, this is probably an unnecessary level of safety - we only use this in one place, and we could easily just pass in a list of existing functions instead of taking them from context.

I'm gonna leave this for now in this PR, and then open another PR to do some of these cleanups

File: src/gcp/cloudfunctions.ts (hunk 6 above)
Comment (joehan) [reply src/gcp/cloudfunctions.ts]: Good point, I moved this into a helper method and changed the signature of the exported method. We currently don't use listFunctions anywhere, but it seems like a usual building block to have around, so i'm leaving it for now

File: src/deploy/functions/prompts.ts (hunk 3 above)
Comment (joehan) [reply src/deploy/functions/prompts.ts]: Ended up being simpler than expected to just pull this call out entirely, so I did so in this pr

Comment (inlined) [reply src/deploy/functions/prompts.ts]: ::sigh:: If I had a dollar for every time I've had to re-edit prompts.ts while doing this refactor. I'm not sure if it's a code smell that this particular file/test breaks so much.
